subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/junit/result/JUnitXmlResultWriter.java

```diff
@@ -83,44 +98,313 @@ private void writeOutputs(SimpleXmlWriter writer, long classId, boolean allClass
+            this.type = type;
+        }
+
+        public void write(SimpleXmlWriter writer) throws IOException {
+            writer.startElement(type.elementName)
+                .attribute("message", failure.getMessage())
+                .attribute("type", failure.getExceptionType());
+
+            if (type.useStacktraceElementAndNestedOutput) {
+                writer.startElement("stackTrace")
+                    .characters(failure.getStackTrace())
+                    .endElement();
+                writeOutput(writer);
+                writer.endElement();
+            } else {
+                writer.characters(failure.getStackTrace());
+                writer.endElement();
+                writeOutput(writer);
+            }
+        }
+    }
+
+    private TestCaseExecution success(long classId, long id) {
+        return new TestCaseExecutionSuccess(outputProvider(classId, id));
+    }
+
+    private TestCaseExecution skipped(long classId, long id) {
+        return new TestCaseExecutionSkipped(outputProvider(classId, id));
+    }
+
+    private Iterable<TestCaseExecution> failures(final long classId, final TestMethodResult methodResult, final FailureType failureType) {
+        List<TestFailure> failures = methodResult.getFailures();
+        final TestFailure firstFailure = failures.get(0);
+        return Iterables.transform(failures, new Function<TestFailure, TestCaseExecution>() {
+            @Override
+            public TestCaseExecution apply(final TestFailure failure) {
+                boolean isFirst = failure == firstFailure;
+                OutputProvider outputProvider = isFirst ? outputProvider(classId, methodResult.getId()) : NullOutputProvider.INSTANCE;
+                return new TestCaseExecutionFailure(outputProvider, failureType, failure);
+            }
+        });
+    }
+
+    private OutputProvider outputProvider(long classId, long id) {
+        return options.outputPerTestCase ? new BackedOutputProvider(classId, id) : NullOutputProvider.INSTANCE;
+    }
+
+    interface OutputProvider {
+        boolean has(TestOutputEvent.Destination destination);
+
+        void write(TestOutputEvent.Destination destination, Writer writer);
+    }
+
+    class BackedOutputProvider implements OutputProvider {
+        private final long classId;
+        private final long testId;
+
+        public BackedOutputProvider(long classId, long testId) {
+            this.classId = classId;
+            this.testId = testId;
+        }
+
+        @Override
+        public boolean has(TestOutputEvent.Destination destination) {
+            return testResultsProvider.hasOutput(classId, testId, destination);
+        }
+
+        @Override
+        public void write(TestOutputEvent.Destination destination, Writer writer) {
+            testResultsProvider.writeTestOutput(classId, testId, destination, writer);
+        }
+    }
+
+    static class NullOutputProvider implements OutputProvider {
```

Comment:
Maybe `enum`?

Reply:
I don't think that's better, as it's an abuse of enum just to achieve a singleton.